File: backend/src/main/java/com/decmoe47/todo/service/impl/TodoServiceImpl.java
package com.decmoe47.todo.service.impl;

import cn.hutool.core.bean.BeanUtil;
import cn.hutool.core.collection.ListUtil;
import com.decmoe47.todo.constant.TodoConstants;
import com.decmoe47.todo.constant.enums.ErrorCodeEnum;
import com.decmoe47.todo.exception.ErrorResponseException;
import com.decmoe47.todo.model.dto.*;
import com.decmoe47.todo.model.entity.Todo;
import com.decmoe47.todo.model.entity.TodoList;
import com.decmoe47.todo.model.entity.User;
import com.decmoe47.todo.model.vo.TodoVO;
import com.decmoe47.todo.repository.TodoListRepository;
import com.decmoe47.todo.repository.TodoRepository;
import com.decmoe47.todo.repository.UserRepository;
import com.decmoe47.todo.service.InboxCacheService;
import com.decmoe47.todo.service.TodoService;
import com.decmoe47.todo.util.SecurityUtil;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import org.springframework.cache.Cache;
import org.springframework.cache.CacheManager;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;

@RequiredArgsConstructor
@Service
public class TodoServiceImpl implements TodoService {

    private final TodoRepository todoRepo;
    private final TodoListRepository todoListRepo;
    private final UserRepository userRepo;
    private final InboxCacheService inboxCacheService;
    private final CacheManager cacheManager;

    @Override
    public List<TodoVO> getTodos(String listId) {
        long userId = SecurityUtil.getCurrentUserId();
        List<Todo> todos;
        if (TodoConstants.INBOX.equals(listId)) {
            listId = inboxCacheService.getInboxId(userId);
        }
        todos = todoRepo.findByCreatedByIdAndBelongedList_Id(userId, listId);
        return BeanUtil.copyToList(todos, TodoVO.class);
    }

    @Override
    public TodoVO addTodo(TodoAddDTO todoAddDTO) {
        User user = userRepo.findById(SecurityUtil.getCurrentUserId())
                .orElseThrow(() -> new ErrorResponseException(ErrorCodeEnum.USER_NOT_FOUND));
        Todo todo = BeanUtil.toBean(todoAddDTO, Todo.class);

        String listId = TodoConstants.INBOX.equals(todoAddDTO.getBelongedListId())
                ? inboxCacheService.getInboxId(user.getId())
                : todoAddDTO.getBelongedListId();
        TodoList todoList = todoListRepo.findById(listId)
                .orElseThrow(() -> new ErrorResponseException(ErrorCodeEnum.TODO_LIST_NOT_FOUND));
        todo.setBelongedList(todoList);

        todo.setCreatedBy(user);
        todo = todoRepo.save(todo);
        return BeanUtil.toBean(todo, TodoVO.class);
    }

    @Transactional
    @Override
    public void deleteTodos(List<TodoDeleteDTO> todoDeleteDTOs) {
        for (TodoDeleteDTO todoDeleteDTO : todoDeleteDTOs) {
            if (Boolean.TRUE.equals(todoDeleteDTO.getSoftDeleted())) {
                todoRepo.softDeleteById(todoDeleteDTO.getId());
            } else {
                todoRepo.deleteById(todoDeleteDTO.getId());
            }
            cleaCache(todoDeleteDTO.getId());
        }
    }

    @Override
    public List<TodoVO> updateTodos(List<TodoUpdateDTO> todoUpdateDTOs) {
        List<Todo> todosToSave = new ArrayList<>();
        for (TodoUpdateDTO todoUpdateDTO : todoUpdateDTOs) {
            Todo todo = todoRepo.findById(todoUpdateDTO.getId())
                    .orElseThrow(() -> new ErrorResponseException(ErrorCodeEnum.TODO_NOT_FOUND));
            BeanUtil.copyProperties(todoUpdateDTO, todo);
            todosToSave.add(todo);
        }

        if (!todosToSave.isEmpty()) {
            todoRepo.saveAll(todosToSave);
            return BeanUtil.copyToList(todosToSave, TodoVO.class);
        } else {
            return ListUtil.empty();
        }
    }

    @Override
    public TodoVO toggleTodo(TodoToggleDTO todoToggleDTO) {
        Todo todo = todoRepo.findById(todoToggleDTO.getId())
                .orElseThrow(() -> new ErrorResponseException(ErrorCodeEnum.TODO_NOT_FOUND));
        todo.setDone(!todo.isDone());
        todo = todoRepo.save(todo);
        return BeanUtil.toBean(todo, TodoVO.class);
    }

    private void cleaCache(long todoId) {
        Cache todoAccess = cacheManager.getCache("todoAccess");
        if (todoAccess != null) {
            todoAccess.evict(SecurityUtil.getCurrentUserId() + "_" + todoId);
        }
    }

    @Override
    public List<TodoVO> moveTodos(List<TodoMoveDTO> todoMoveDTOs) {
        List<Todo> todosToSave = new ArrayList<>();
        for (TodoMoveDTO todoMoveDTO : todoMoveDTOs) {
            Todo todo = todoRepo.findById(todoMoveDTO.getId())
                    .orElseThrow(() -> new ErrorResponseException(ErrorCodeEnum.TODO_NOT_FOUND));

            String listId = TodoConstants.INBOX.equals(todoMoveDTO.getTargetListId())
                    ? inboxCacheService.getInboxId(SecurityUtil.getCurrentUserId())
                    : todoMoveDTO.getTargetListId();
            TodoList todoList = todoListRepo.findById(listId)
                    .orElseThrow(() -> new ErrorResponseException(ErrorCodeEnum.TODO_LIST_NOT_FOUND));      // TODO: 改用mybatis plus直接update

            todo.setBelongedList(todoList);
            todosToSave.add(todo);
        }

        if (!todosToSave.isEmpty()) {
            todoRepo.saveAll(todosToSave);
            return BeanUtil.copyToList(todosToSave, TodoVO.class);
        } else {
            return ListUtil.empty();
        }
    }
}

File: backend/src/main/kotlin/com/decmoe47/todo/service/impl/TodoServiceImpl.kt
package com.decmoe47.todo.service.impl

import com.decmoe47.todo.annotation.ReadOnlyTransactionalService
import com.decmoe47.todo.constant.enums.ErrorCode
import com.decmoe47.todo.exception.ErrorResponseException
import com.decmoe47.todo.model.entity.AuditableEntity
import com.decmoe47.todo.model.entity.Todo
import com.decmoe47.todo.model.mapper.toTodoResponse
import com.decmoe47.todo.model.request.*
import com.decmoe47.todo.model.response.TodoResponse
import com.decmoe47.todo.repository.TodoListRepository
import com.decmoe47.todo.repository.TodoRepository
import com.decmoe47.todo.repository.UserRepository
import com.decmoe47.todo.service.TodoService
import com.decmoe47.todo.util.SecurityUtil
import org.springframework.cache.CacheManager
import org.springframework.transaction.annotation.Transactional

@ReadOnlyTransactionalService
class TodoServiceImpl(
    private val todoRepo: TodoRepository,
    private val todoListRepo: TodoListRepository,
    private val userRepo: UserRepository,
    private val cacheManager: CacheManager,
) : TodoService {
    override fun getTodos(listId: Long): List<TodoResponse> {
        val todos = todoRepo.select(listId)
        return todos.map { it.toTodoResponse() }
    }

    @Transactional
    override fun addTodo(request: TodoAddRequest): TodoResponse {
        val userId = SecurityUtil.getCurrentUserId()
        val user = userRepo.first(userId) ?: throw ErrorResponseException(ErrorCode.USER_NOT_FOUND)
        val todoList = todoListRepo.first(request.belongedListId)
            ?: throw ErrorResponseException(ErrorCode.TODO_LIST_NOT_FOUND)

        val todo = Todo(
            content = request.content,
            dueDate = request.dueDate,
            belongedListId = todoList.id,
            auditable = AuditableEntity(createdBy = user.id),
        )
        return todoRepo.save(todo).toTodoResponse()
    }

    @Transactional
    override fun deleteTodo(request: TodoDeleteRequest) {
        if (request.softDeleted) {
            todoRepo.softDelete(request.id)
        } else {
            todoRepo.delete(request.id)
        }
        clearCache(request.id)
    }

    @Transactional
    override fun updateTodo(request: TodoUpdateRequest): TodoResponse {
        val todo = todoRepo.first(request.id)
            ?: throw ErrorResponseException(ErrorCode.TODO_NOT_FOUND)
        val updatedTodo = todo.copy(
            content = request.content,
            done = request.done,
            dueDate = request.dueDate,
            description = request.description,
            auditable = todo.auditable.copy(updatedBy = SecurityUtil.getCurrentUserId()),
        )
        return todoRepo.update(updatedTodo).toTodoResponse()
    }

    @Transactional
    override fun toggleTodo(request: TodoToggleRequest): TodoResponse {
        val userId = SecurityUtil.getCurrentUserId()
        val todo = todoRepo.first(request.id)
            ?: throw ErrorResponseException(ErrorCode.TODO_NOT_FOUND)

        val updatedTodo = todoRepo.update(
            todo.copy(
                done = !todo.done,
                auditable = todo.auditable.copy(updatedBy = userId),
            )
        )
        return todoRepo.update(updatedTodo).toTodoResponse()
    }

    @Transactional
    override fun moveTodo(request: TodoMoveRequest): TodoResponse {
        val userId = SecurityUtil.getCurrentUserId()

        val todo = todoRepo.first(request.id)
            ?: throw ErrorResponseException(ErrorCode.TODO_NOT_FOUND)

        val todoList = todoListRepo.first(request.targetListId)
            ?: throw ErrorResponseException(ErrorCode.TODO_LIST_NOT_FOUND)

        val updatedTodo = todo.copy(
            belongedListId = todoList.id,
            auditable = todo.auditable.copy(updatedBy = userId),
        )

        return todoRepo.update(updatedTodo).toTodoResponse()
    }

    private fun clearCache(todoId: Long) {
        val todoAccess = cacheManager.getCache("todoAccess")
        todoAccess?.evict("${SecurityUtil.getCurrentUserId()}_${todoId}")
    }
}
